GUIDE.md

```diff
@@ -0,0 +1,299 @@
+
+### Chain
+
+Chain is a handler which allows to execute more than one handler:
+
+```rust
+use carapax::Chain;
+
+async fn handler1(_: ()) {}
+async fn handler2(_: ()) {}
+async fn handler3(_: ()) {}
+
+let chain = Chain::default()
+    .add_handler(handler1)
+    .add_handler(handler2)
+    .add_handler(handler3);
+```
+
+Handlers will run in same order as added.
+If a handler returns `HandlerResult::Stop` or `HandlerResult::Error(_)`, all handlers added after that handler will not run.
+
+### HandlerResult
+
+Normally every handler must return a `HandlerResult` or any type which converts into it:
+
+| From            | To                          |
+|-----------------|-----------------------------|
+| `()`            | `HandlerResult::Stop`       |
+| `true`          | `HandlerResult::Continue`   |
+| `false`         | `HandlerResult::Stop`       |
+| `Result<T, E>`  | `T.into::<HandlerResult>()` |
+
+
+## Predicates
+
+[`Predicate`](https://tg-rs.github.io/carapax/carapax/struct.Predicate.html)
+is a decorator which allows to decide, should handler run or not.
+It is useful when you need to implement a ratelimit or close a handler or entire bot from unwanted users and so on.
+
+A predicate handler must implement `Handler` trait, but return
+a [`PredicateResult`](https://tg-rs.github.io/carapax/carapax/enum.PredicateResult.html)
+or any type which converts into it:
+
+|   From            | To                            |
+|-------------------|-------------------------------|
+| `true`            | `PredicateResult::Continue`   |
+| `false`           | `PredicateResult::Stop`       |
+| `Result<T, E>`    | `T.into::<PredicateResult>()` |
+
+
+Example:
+
+```rust
+use carapax::{
+    methods::SendMessage,
+    types::{ChatId, Text},
+    Api, Chain, PredicateExt, Ref,
+};
+
+fn setup(chain: &mut Chain) {
+    let handler = pong.predicate(is_ping);
+    chain.add_handler(handler);
+}
+
+async fn is_ping(text: Text) -> bool {
+    text.data == "ping"
+}
+
+async fn pong(api: Ref<Api>, chat_id: ChatId) {
+    let method = SendMessage::new(chat_id, "pong");
+    api.execute(method).await.unwrap();
+}
+
+```
+
+### Commands
+
+[`CommandPredicate`](https://tg-rs.github.io/carapax/carapax/struct.CommandPredicate.html#)
+is a decorator which allows to run handlers only when user sent a command.
+Note that command name contains a leading slash (`/`).
+
+```rust
+use carapax::{
+    methods::SendMessage,
+    types::{ChatId, User},
+    Api, Chain, CommandExt, Ref,
+};
+
+fn setup(chain: &mut Chain) {
+    let handler = greet.command("/hello");
+    chain.add_handler(handler);
+}
+
+async fn greet(api: Ref<Api>, chat_id: ChatId, user: User) {
+    let method = SendMessage::new(chat_id, format!("Hello, {}", user.first_name));
+    api.execute(method).await.unwrap();
+}
+
+```
+
+### Access
+
+[AccessPredicate](https://tg-rs.github.io/carapax/carapax/access/struct.AccessPredicate.html) allows to protect handlers from unwanted users.
```

Comment:
```suggestion
[AccessPredicate](https://tg-rs.github.io/carapax/carapax/access/struct.AccessPredicate.html) allows to prevent a handler from being called if a condition is false.
```

Reply:
Condition?